Update dependency installs and prepare for pip deployment

rmkraus opened the pull request:
In this PR, I will be updating PIP dependency installations to install to config directory rather than the Python library. I will also be getting Home Assistant ready to be deployed using PIP rather than GIT. The PIP deployment will also include an executable.

rmkraus committed:
Created setup.py for standard installs.

rmkraus commented:
No need to pull yet. I created the PR to watch the updates.

rmkraus committed:
Cleaned up main file to remove dependency management.

rmkraus committed:
Many changes to cleanup config directory and lib installations.

Cleaned up default config directory determination.
Made bootstrap creators for HA always set config directory.
Made bootstrap creators set the local library in the Python Path.
Moved all exceptions to their own file to make imports easier.
Moved default configuration directory be in the users’ profile.
Moved pip installs to be done to a lib folder in the config directory.
Reduced requirements.txt to only the barebones reqs.

balloob commented on homeassistant/__main__.py:
```diff
@@ -31,54 +30,6 @@ def ensure_pip():
-    return bootstrap
-
-
 def validate_git_submodules():
```
We can remove this method too.

balloob commented on homeassistant/__main__.py:
```diff
@@ -94,13 +45,25 @@ def ensure_config_path(config_dir):
-
-    import homeassistant.config as config_util
+        try:
+            os.mkdir(config_dir)
```
We should only try to make it if we are using the default one. Imaging making a typo and Home Assistant is creating a second dir !

balloob commented on homeassistant/__main__.py:
```diff
@@ -149,21 +106,18 @@ def main():
-        from homeassistant.components import frontend, demo
-
         hass = bootstrap.from_config_dict({
             frontend.DOMAIN: {},
```
Might as well just write `frontend` and `demo` as strings. Don't know why I ever bothered using `DOMAIN`

balloob commented on homeassistant/core.py:
```diff
@@ -660,7 +664,11 @@ def __init__(self):
-        self.config_dir = os.path.join(os.getcwd(), 'config')
+        self.config_dir = get_default_config_dir()
+
+    def mount_local_path(self):
```
This should not be a concern of the HA core. Let's move it to bootstrap?

balloob commented on setup.py:
```diff
@@ -0,0 +1,47 @@
+    version=VERSION,
+    license='MIT License',
+    url='https://home-assistant.io/',
+    download_url='https://github.com/automicus/pyisy/tarball/0.7.0',
```
This should be updated.

rmkraus replied:
Haha. Copy Pasta strikes again.

balloob commented on setup.py:
```diff
@@ -0,0 +1,47 @@
+    package_data=PACKAGE_DATA,
+    zip_safe=False,
+    platforms='any',
+    install_requires=['requests==2.7.0', 'pyyaml==3.11', 'pytz==2015.4'],
```
Maybe we can be a bit more flexible here. Like `requests>2,<3`, `pyyaml>3.11,<4`, `pytz>=2015.4`

For pytz I am not afraid that it will add breaking changes in future releases

balloob commented on setup.py:
```diff
@@ -0,0 +1,47 @@
+        ]
+    },
+    classifiers=[
+        'Intended Audience :: End Users/Desktop',
```
Should we add `Intended Audience :: Developers` because you can write Python code to interact with HA? Or is that tag for lib only

rmkraus replied:
I don't think there are any hard and fast rules. I think we can add both.

rmkraus replied:
Done.

balloob commented on requirements.txt:
```diff
@@ -1,119 +1,3 @@
-# Required for Home Assistant core
-requests>=2.0
```
Would it be possible that we keep `requirements.txt` around but just for the core dependencies. Have `setup.py` read the dependencies in from this file for the core.

Move all platform dependencies to `requirements_all.txt`. That way I can still have Docker install them all.

balloob replied:
Also add `pip3 install -r requirements_all.txt` to our `.travis.yml` file to solve the test issues.

rmkraus replied:
Done.

rmkraus committed:
Changed component REQUIREMENTS to absolute versions.

rmkraus committed:
Implemented comments from Paulus.

Revised main to use frontend and demo strings rather than importing
their domains.
Removed submodule validation.
Moved local library mounting to the bootstrap module and out of core.
Added requirements_all.txt for all dependencies.
Made core dependencies looser.
Small updates to setup.py.

rmkraus committed:
Mostly PyLint and Flake8 updates.

Rewrote imports of exceptions to be from the exceptions module.
Made nmap scanner check for libnmap dependency without crashing.
Various flake8 and pylint updates.

rmkraus committed:
Updated core config directory tests

rmkraus commented:
Still have to check various versions of pip.

balloob commented on homeassistant/__main__.py:
```diff
@@ -31,76 +29,34 @@ def ensure_pip():
-        sys.exit()
-
-    import homeassistant.config as config_util
+        if config_dir == config_util.get_default_config_dir():
```
Not a comment to fix but more a tip for the future. This is a perfect place to use a guard clause to reduce code complexity:

```python
if not os.path.isdir(config_dir):
    if config_dir != config_util.get_default_config_dir():
        print(('Fatal Error: Specified configuration directory does '
               'not exist {} ').format(config_dir))
        sys.exit()
    # Other code that previously was in an else but now not anymore
```

balloob commented on homeassistant/__main__.py:
```diff
@@ -31,76 +29,34 @@ def ensure_pip():
+            except OSError:
+                print(('Fatal Error: Unable to create default configuration '
+                       'directory {} ').format(config_dir))
+                sys.exit()
```
I know this was already here but let's change this to `sys.exit(1)` to show the system that an error caused Home Assistant to quit.

balloob commented on homeassistant/__main__.py:
```diff
@@ -149,24 +97,20 @@ def main():
                 import webbrowser
                 webbrowser.open(hass.config.api.base_url)
-
+        from homeassistant.const import EVENT_HOMEASSISTANT_START
```
this one can be moved to the top of the file too

balloob commented:
Do we need these files?
 - `homeassistant/components/frontend/www_static/__init__.py`
 - `homeassistant/components/frontend/www_static/images/__init__.py`

rmkraus committed:
Pip requirements and fixes.

1) Reduced the flags being sent to pip.
2) Required a minimum of pip 7.0.0 for Home Assistant.

rmkraus commented:
Yeah, we do. setuptools only like to work with packages. There might be other ways to do it. I tried a few things and this one is the one I got to work the best.

rmkraus committed:
Minor tweaks

Moved another import to the top of main.
Forced an exit code of 1 when there is an error initializing.

rmkraus commented:
One last check. I need to verify that everything plays nicely with system level Python installs.

rmkraus commented:
Works on my Ubuntu server without a virtual environment. Ready for pulling pending further review.
